### PR TITLE
fix: Error on invalid external types

### LINKED ISSUE #34: Returning sensible error messages about types to user
As brought up in #18, when users supply the wrong type to the data - for example, giving the type int to a date - we want to be able to loudly return this typing error back to the user. 

The function `type_variables_in_df()` in `cohortreport.processing` does not raise any errors if it is not able to type correctly. I believe this is where it might be sensible to return something to the user.

### PR DESCRIPTION
**Pull request by iaindillingham**
When a user passes an "untyped" input file (`.csv` and `.csv.gz`) to cohort-report, they must also pass a configuration object called `variable_types` that maps from names to _external_ types (e.g. `binary`, `categorical`, etc.). Previously, invalid external types that were valid _internal_ types (i.e. Pandas types) were permitted; invalid external types that were invalid internal types were not.

This PR ensures that cohort-report now errors on all invalid external types. The errors are also more user-friendly. The commits are in three groups:

* Refactor and expand the tests (7717630 to 028ae06)
* Fix the issue (4d55563)
* Refactor (589c74e to d67bac2)

BREAKING CHANGE: Cohort-report now errors on all invalid external types.

Fixes #34

### CODE REVIEW

File: tests/test_processing.py [hunk 1]
```diff
@@ -68,23 +68,24 @@ def test_columns_match(self, test_df):
...
-                ],
-                "test_float": [1.2, 5.4],
-            }
-        )
+@pytest.fixture
+def test_df():
+    return pd.DataFrame(
+        {
+            "patient_id": [1, 2],
+            "test_binary": [1, 0],
```
Comment (CarolineMorton) [inline tests/test_processing.py:56]: agree this is much easier to read

File: tests/test_processing.py [hunk 2]
```diff
@@ -121,12 +121,8 @@ def test_external_types_do_not_map_to_internal_types_and_are_valid_internal_type
...
             "test_float": "string",
         }
-        typed_df = processing.type_variables_in_df(input_dataframe, variable_types)
-        assert typed_df["test_binary"].dtype == "string"
-        assert typed_df["test_categorical"].dtype == "string"
-        assert typed_df["test_int"].dtype == "string"
-        assert typed_df["test_date"].dtype == "string"
-        assert typed_df["test_float"].dtype == "string"
+        with pytest.raises(ValueError):
+            processing.type_variables_in_df(input_dataframe, variable_types)
```
Comment (CarolineMorton) [inline tests/test_processing.py]: maybe check the value of the ValueError

**Review by CarolineMorton — APPROVED** [review]
looks good

one comment

File: tests/test_processing.py (hunk 2 above)
Comment (iaindillingham) [reply tests/test_processing.py]: Done in 01ecadd 👍🏻

**Pull request merged**